azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -0,0 +1,151 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+""" This module contains the manager for handler methods used by the aio clients"""
+
+from azure.iot.device.common import asyncio_compat
+from azure.iot.device.common.chainable_exception import ChainableException
+from concurrent.futures import ThreadPoolExecutor
+import asyncio
+import logging
+import inspect
+
+logger = logging.getLogger(__name__)
+
+MESSAGE = "_on_message_received"
+METHOD = "_on_method_request_received"
+TWIN_DP_PATCH = "_on_twin_desired_properties_patch_received"
+# TODO: add more for "event"
+
+
+class AsyncHandlerManagerException(ChainableException):
+    pass
+
+
+class AsyncHandlerManager(object):
+    def __init__(self, inbox_manager):
+        self._inbox_manager = inbox_manager
+
+        self._handler_tasks = {
+            # Inbox handler tasks
+            MESSAGE: None,
+            METHOD: None,
+            TWIN_DP_PATCH: None,
+            # Other handler tasks
+            # TODO: add
+        }
+
+        # Inbox handlers
+        self._on_message_received = None
+        self._on_method_request_received = None
+        self._on_twin_desired_properties_patch_received = None
+
+        # Other handlers
+        # TODO: add
+
+    async def _run_inbox_handler(self, inbox, handler_name):
+        """Run infinite loop that waits for an inbox to receive an object from it, then calls
+        the handler with that object
+        """
+        # Run the handler in a threadpool, so that it cannot block other handlers (from a different task),
```

Comment:
># Run the handler in a threadpool, so that it cannot block other handlers (from a different task), [](start = 8, length = 98)

I'm having a hard time convincing myself that this is the right decision.  mixing threads and coroutines seems dangerous one of the big advantages of coroutines is that you don't have to worry about multiple threads so you don't have to worry about synchronizing those threads.  you're breaking that assumption and I don't know if that is cool.

I read (part of) a thread on this.  I'll forward it because there's a lot of good info & discussions in it.

Reply:
that being said, I don't think this needs to block your PR.  This particular decision can always change because the rest of your code isn't tightly bound to this.    

---
In reply to: [454548409](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r454548409) [](ancestors = 454548409)

Reply:
I also tried to avoid this, but in order to get the performance we want, for various reasons, it has to be this way. Let's discuss further on a call.

---
In reply to: [454549029](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r454549029) [](ancestors = 454549029,454548409)

Reply:
We have discussed to potentially change this design in a future PR that address widespread client performance issues related to customer event loops

---
In reply to: [454567833](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r454567833) [](ancestors = 454567833,454549029,454548409)